File: src/main/java/com/thunderbear06/client/EmissiveTextureLayer.java
package com.thunderbear06.client;

import com.mojang.blaze3d.vertex.PoseStack;
import com.mojang.blaze3d.vertex.VertexConsumer;
import java.util.function.Function;
import net.minecraft.client.model.EntityModel;
import net.minecraft.client.renderer.MultiBufferSource;
import net.minecraft.client.renderer.RenderType;
import net.minecraft.client.renderer.entity.RenderLayerParent;
import net.minecraft.client.renderer.entity.layers.RenderLayer;
import net.minecraft.client.renderer.texture.OverlayTexture;
import net.minecraft.resources.ResourceLocation;
import net.minecraft.world.entity.Entity;

final class EmissiveTextureLayer<T extends Entity, M extends EntityModel<T>> extends RenderLayer<T, M> {
    private final Function<T, ResourceLocation> texture;

    EmissiveTextureLayer(RenderLayerParent<T, M> parent, Function<T, ResourceLocation> texture) {
        super(parent);
        this.texture = texture;
    }

    @Override
    public void render(PoseStack poseStack, MultiBufferSource buffer, int packedLight, T entity,
                       float limbSwing, float limbSwingAmount, float partialTicks, float ageInTicks,
                       float netHeadYaw, float headPitch) {
        ResourceLocation location = texture.apply(entity);
        if (location == null) return;
        VertexConsumer vertices = buffer.getBuffer(RenderType.entityTranslucentEmissive(location));
        getParentModel().renderToBuffer(poseStack, vertices, 0xF000F0, OverlayTexture.NO_OVERLAY);
    }
}

File: src/main/kotlin/com/thunderbear06/client/EmissiveTextureLayer.kt
package com.thunderbear06.client

import com.mojang.blaze3d.vertex.PoseStack
import net.minecraft.client.model.EntityModel
import net.minecraft.client.renderer.MultiBufferSource
import net.minecraft.client.renderer.RenderType
import net.minecraft.client.renderer.entity.RenderLayerParent
import net.minecraft.client.renderer.entity.layers.RenderLayer
import net.minecraft.client.renderer.texture.OverlayTexture
import net.minecraft.resources.ResourceLocation
import net.minecraft.world.entity.Entity

internal class EmissiveTextureLayer<T : Entity, M : EntityModel<T>>(
    parent: RenderLayerParent<T, M>,
    private val texture: (T) -> ResourceLocation?,
) : RenderLayer<T, M>(parent) {
    override fun render(
        poseStack: PoseStack,
        buffer: MultiBufferSource,
        packedLight: Int,
        entity: T,
        limbSwing: Float,
        limbSwingAmount: Float,
        partialTicks: Float,
        ageInTicks: Float,
        netHeadYaw: Float,
        headPitch: Float,
    ) {
        val location = texture(entity) ?: return
        val vertices = buffer.getBuffer(RenderType.entityTranslucentEmissive(location))
        parentModel.renderToBuffer(poseStack, vertices, 0xF000F0, OverlayTexture.NO_OVERLAY)
    }
}
